# Motivation

I was never trained to be a good programmer. I somehow passed the programming exam in my education and did not learn too much. I wanted to concque this challenage once for all to prove myself: no matter how low one's natual ability is, one can conquer anything with hardwork and persistent and it is never too late!

[Reference](https://www.techinterviewhandbook.org/grind75/faq)

# Week 1

##  (1) Two Sum

Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

 

Example 1:
```
Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
```
Example 2:

```
Input: nums = [3,2,4], target = 6
Output: [1,2]
```

Example 3:

```
Input: nums = [3,3], target = 6
Output: [0,1]
```
 
Constraints:

* 2 <= nums.length <= 104
* -109 <= nums[i] <= 109
* -109 <= target <= 109
* Only one valid answer exists.

Follow-up: Can you come up with an algorithm that is less than O(n^2) time complexity?

Using hashtable can get 
* time: $O(n)$
* space: $O(n)$

**key takeaway**

Use hashtable to store sum subtract current item to get desired item as we iterate the list to improve time complexity

In [1]:
from typing import List, Optional

class Solution:
    def twoSum1(self, nums: List[int], target: int) -> List[int]:
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                num1 = nums[i]
                num2 = nums[j]
                if num2 == target - num1:
                    return [i, j]
    
    def twoSum2(self, nums, target):
        lookup = {}
        for i, num in enumerate(nums):
            if target - num in lookup:
                return [lookup[target - num], i]
            lookup[num] = i

s = Solution()

%timeit -n5 s.twoSum1([1, 3, 4, 8, 10, 2,7,11,15], 22)
%timeit -n5 s.twoSum2([1, 3, 4, 8, 10, 2,7,11,15], 22)

6.19 µs ± 257 ns per loop (mean ± std. dev. of 7 runs, 5 loops each)
1.56 µs ± 267 ns per loop (mean ± std. dev. of 7 runs, 5 loops each)


## (20) Valid Parentheses

Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Every close bracket has a corresponding open bracket of the same type.
 

Example 1:

```
Input: s = "()"
Output: true
```

Example 2:
```
Input: s = "()[]{}"
Output: true
```
Example 3:
```
Input: s = "(]"
Output: false
``` 

Constraints:

* 1 <= s.length <= 104
* s consists of parentheses only '()[]{}'.

Big O
* Time $O(n)$
* Space $O(n)$

**Key takeaaway**

Use push open parenthese to a stack, pop if matching, at the end if stack is empty, return true

In [8]:
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) <= 1:
            return False
        
        stack = []
        for c in s:
            if c == "(" or c == "[" or c == "{":
                stack.append(c)
            elif c == ")":
                if len(stack) == 0 or stack.pop(-1) != "(":
                    return False
            elif c == "}":
                if len(stack) == 0 or stack.pop(-1) != "{":
                    return False
            elif c == "]":
                if len(stack) == 0 or stack.pop(-1) != "[":
                    return False
        return len(stack) == 0

## (21) Merge Two Sorted Lists

You are given the heads of two sorted linked lists list1 and list2.

Merge the two lists in a one sorted list. The list should be made by splicing together the nodes of the first two lists.

Return the head of the merged linked list.

Example 1:

```
Input: list1 = [1,2,4], list2 = [1,3,4]
Output: [1,1,2,3,4,4]
```
Example 2:

```
Input: list1 = [], list2 = []
Output: []
```
Example 3:
```
Input: list1 = [], list2 = [0]
Output: [0]
```

Constraints:

* The number of nodes in both lists is in the range [0, 50].
* -100 <= Node.val <= 100
* Both list1 and list2 are sorted in non-decreasing order.

Iterative:
* time: $O(n+m)$
* space: $O(1)$

Recursive:
* time: $O(n+m)$
* space: $O(n+m)$

**key takeaway**
either use recursive as merge list with space cost or just insert each node from one list to the other

In [39]:
#Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists_recursive(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if list1 is None:
            return list2
        if list2 is None:
            return list1
        
        if list1.val < list2.val:
            list1.next = self.mergeTwoList(list1.next, list2)
            return list1
        else:
            list2.next = self.mergeTwoLists(list1, list2.next)
            return list2
        
        
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        new_list = ListNode()
        head = new_list
        tail = new_list
        
        while list1 is not None and list2 is not None:
            if list1.val < list2.val:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            
            tail = tail.next
        
        if list1 is None:
            tail.next = list2
        else:
            tail.next = list1
            
        return head.next
            
list1 = ListNode(1, ListNode(2, ListNode(4))) 
list2 = ListNode(1, ListNode(3, ListNode(4))) 

s = Solution()
n = s.mergeTwoLists_recursive(list1, list2)
while n is not None:
    print(n.val)
    n = n.next

1
1
2
3
4
4


## (121) Best Time to buy and sell stock

You are given an array prices where prices[i] is the price of a given stock on the ith day.

You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.

 

Example 1:
```
Input: prices = [7,1,5,3,6,4]
Output: 5
Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
Note that buying on day 2 and selling on day 1 is not allowed because you must buy before you sell.
```
Example 2:
```
Input: prices = [7,6,4,3,1]
Output: 0
Explanation: In this case, no transactions are done and the max profit = 0.
```
 
Constraints:

* 1 <= prices.length <= 105
* 0 <= prices[i] <= 104

Big O
* Time $O(n)$
* Space $O(1)$

**key takeway**
Use a 2 pointer left and right (sliding window) update the pointer based on the local min in the window, update max profit based on the local max of the window

In [56]:
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) <= 1:
            return 0
        l, r = 0, 1
        maxprofit = 0
        while r < len(prices):
            if prices[r] > prices[l]:
                profit = prices[r] - prices[l]
                maxprofit = max(maxprofit, profit)
            else:
                l = r
            r+=1
        return maxprofit
    
s = Solution()
print(s.maxProfit([7,1,5,3,6,4]))
print(s.maxProfit([7,6,4,3,1]))
%timeit -n5 s.maxProfit([7,1,5,3,6,4])
%timeit -n5 s.maxProfit([7,6,4,3,1])

5
0
3.94 µs ± 332 ns per loop (mean ± std. dev. of 7 runs, 5 loops each)
1.97 µs ± 206 ns per loop (mean ± std. dev. of 7 runs, 5 loops each)


## (125) Valid Palindrome


A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.

Given a string s, return true if it is a palindrome, or false otherwise.

Example 1:

```
Input: s = "A man, a plan, a canal: Panama"
Output: true
Explanation: "amanaplanacanalpanama" is a palindrome.
```

Example 2:

```
Input: s = "race a car"
Output: false
Explanation: "raceacar" is not a palindrome.
```
Example 3:

```
Input: s = " "
Output: true
Explanation: s is an empty string "" after removing non-alphanumeric characters.
Since an empty string reads the same forward and backward, it is a palindrome.
```
 
Constraints:

* 1 <= s.length <= 2 * 105
* s consists only of printable ASCII characters.

Big O
* Time $O(n)$
* Space $O(n)$ to store cleaned string
* Use 2 pointers to get $O(1)$ space

**key takeaway**

Using string slice:
* Python regrex `re.sub(r'[^0-9a-zA-Z]', '', s)` to remove nonalphanumeric
* String reverse using slice `reversed = s[::-1]`
* Palindrome means `s == s[::-1] => True`

User 2 pointers:
* Start left, right pointer at both end
* skip nonalphanumeric char
* convert to lower and compare if not equal return false
* This approach which would save space with $O(1)$

In [75]:
import re

class Solution:
    def isPalindrome(self, s: str) -> bool:
        cleaned_s = re.sub(r'[^0-9a-zA-Z]+', '', s.lower())
        if len(cleaned_s) <= 1:
            return True
        
        if len(cleaned_s)%2 == 0:
            return cleaned_s[:len(cleaned_s)//2] == cleaned_s[-len(cleaned_s)//2:][::-1]
        else:
            return cleaned_s[:len(cleaned_s)//2] == cleaned_s[-len(cleaned_s)//2+1:][::-1]

    def isPalindrome(self, s):
        s_clr = re.sub(r'[^a-zA-Z0-9]', '', s).lower()
        
        if s_clr == s_clr[::-1]:
            return True
        else:
            return False

    def isPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1
        while left < right:
            if not s[left].isalnum():
                left += 1
            elif not s[right].isalnum():
                right -= 1
            elif s[left].lower() != s[right].lower():
                return False
            else:
                left += 1
                right -= 1
        return True

    # if isalnum() is not allowed we need to create one
    def alphaNum(self, c):
        return (ord('A') <= ord(c) <= ord['Z'] or
                ord('a') <= ord(c) <= ord['z'] or
               ord('0') <= ord(c) <= ord['9'] or
        
s = Solution()
s.isPalindrome("A man, a plan, a canal: Panama")

True

## (226) Invert Binary Tree

Given the root of a binary tree, invert the tree, and return its root.

 

Example 1:

```
Input: root = [4,2,7,1,3,6,9]
Output: [4,7,2,9,6,3,1]
```

Example 2:

```
Input: root = [2,1,3]
Output: [2,3,1]
```
Example 3:

```
Input: root = []
Output: []
```
 

Constraints:

* The number of nodes in the tree is in the range [0, 100].
* -100 <= Node.val <= 100

Big O

* Time $O(h)$ height (or $O(log2(n)$
* Space $O(n)$

**Key Takeaway**
Use recursion to perform DFS (understanding of DFS)

In [76]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if root is None:
            return None
        
        temp = root.left
        root.left = root.right
        root.right = temp
        self.invertTree(root.left)
        self.invertTree(root.right)
        return root
    
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None
        
        right = self.invertTree(root.right)
        left = self.invertTree(root.left)
        root.left = right
        root.right = left
        return root

    # iteratively (use Queue)
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None
        
        queue = collections.deque([root])
        while queue:
            current = queue.popleft()
            current.left, current.right = current.right, current.left
            
            if current.left:
                queue.append(current.left)
            
            if current.right:
                queue.append(current.right)
        
        return root

## (242) Valid Anagram

Given two strings s and t, return true if t is an anagram of s, and false otherwise.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.
 

Example 1:

```
Input: s = "anagram", t = "nagaram"
Output: true
```
Example 2:

```
Input: s = "rat", t = "car"
Output: false
```

Constraints:

* 1 <= s.length, t.length <= 5 * 10^4
* s and t consist of lowercase English letters.

Big O:

* If use sort, time: $O(nlogn)$ with space either $O(1)$ or $O(n)$ depending on sort algorithm
* If use hashmap to store count and compare, time: $O(s+t)$, and space $O(s+t)$

**Key Takeaway**
* use hashmap to store count for each char and compare
* check length first to improve performance
* We can also use sort but different big O.

In [ ]:
class Solution:
    # Approach 1: sorted and compare directly
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        
        if sorted(s) == sorted(t):
            return True
        else:
            return False
    
    # Approach 2: sorted and compare in a loop 
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        sorted_s = sorted(s)
        sorted_t = sorted(t)
        for i in range(len(s)):
            if sorted_s[i] != sorted_t[i]:
                return False
        return True
    
    # Count frequency of chars and compare
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        s_dict = {}
        t_dict = {}
        for i in range(len(s)):
            s_dict[s[i]] = 1 + s_dict.get(s[i], 0)
            t_dict[t[i]] = 1 + t_dict.get(t[i], 0)
            
        for c in s_dict:
            if s_dict[c] != t_dict.get(c, 0):
                return False
        return True
    
    # Count frequency of chars and compare
    def isAnagram(self, s: str, t: str) -> bool:
        return Counter(s) == Counter(t)

## (704) Binary Search

Given an array of integers nums which is sorted in ascending order, and an integer target, write a function to search target in nums. If target exists, then return its index. Otherwise, return -1.

You must write an algorithm with O(log n) runtime complexity.

 

Example 1:
```
Input: nums = [-1,0,3,5,9,12], target = 9
Output: 4
Explanation: 9 exists in nums and its index is 4
```
Example 2:
```
Input: nums = [-1,0,3,5,9,12], target = 2
Output: -1
Explanation: 2 does not exist in nums so return -1
```

Constraints:

* 1 <= nums.length <= 104
* -104 < nums[i], target < 104
* All the integers in nums are unique.
* nums is sorted in ascending order.

Big O:

* Time $O(log n)$

In [ ]:
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        size = len(nums)
        low = 0
        high = size - 1
        
        while low <= high:
            mid = int((low + high)/2)
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                high = mid - 1
            else:
                low = mid + 1
        return -1

## (733) Flood Fill

An image is represented by an m x n integer grid image where image[i][j] represents the pixel value of the image.

You are also given three integers sr, sc, and color. You should perform a flood fill on the image starting from the pixel image[sr][sc].

To perform a flood fill, consider the starting pixel, plus any pixels connected 4-directionally to the starting pixel of the same color as the starting pixel, plus any pixels connected 4-directionally to those pixels (also with the same color), and so on. Replace the color of all of the aforementioned pixels with color.

Return the modified image after performing the flood fill.

 

Example 1:

```
Input: image = [[1,1,1],[1,1,0],[1,0,1]], sr = 1, sc = 1, color = 2
Output: [[2,2,2],[2,2,0],[2,0,1]]
Explanation: From the center of the image with position (sr, sc) = (1, 1) (i.e., the red pixel), all pixels connected by a path of the same color as the starting pixel (i.e., the blue pixels) are colored with the new color.
Note the bottom corner is not colored 2, because it is not 4-directionally connected to the starting pixel.
```
Example 2:
```
Input: image = [[0,0,0],[0,0,0]], sr = 0, sc = 0, color = 0
Output: [[0,0,0],[0,0,0]]
Explanation: The starting pixel is already colored 0, so no changes are made to the image.
```

Constraints:

* m == image.length
* n == image[i].length
* 1 <= m, n <= 50
* 0 <= image[i][j], color < 216
* 0 <= sr < m
* 0 <= sc < n

**key takeways**

* A typical DFS problem with recursion. A good way to learn DFS
* The base case is only fill when the color is equal to the current color of initial pixel

In [88]:
class Solution:
    def dfs(self, image, r, c, old_color, new_color):
        if image[r][c] == old_color:
            image[r][c] = new_color
            if r >= 1:
                self.dfs(image, r-1, c, old_color, new_color)
            if r < len(image) - 1:
                self.dfs(image, r+1, c, old_color, new_color)
            if c >= 1:
                self.dfs(image, r, c-1, old_color, new_color)
            if c < len(image[0])-1:
                self.dfs(image, r, c+1, old_color, new_color)
        else:
            return
                
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        old_color = image[sr][sc]
        if old_color == color:
            return image
        else:
            self.dfs(image, sr, sc, old_color, color)
        return image
        
s = Solution()
image = [[1,1,1],[1,1,0],[1,0,1]]
sr = 1
sc = 1
color = 2
print(s.floodFill(image, sr, sc, color))

[[2, 2, 2], [2, 2, 0], [2, 0, 1]]


## (235) Lowest Common Ancestor of a Binary Search Tree

Given a binary search tree (BST), find the lowest common ancestor (LCA) node of two given nodes in the BST.

According to the definition of LCA on Wikipedia: “The lowest common ancestor is defined between two nodes p and q as the lowest node in T that has both p and q as descendants (where we allow a node to be a descendant of itself).”

 

Example 1:

```
Input: root = [6,2,8,0,4,7,9,null,null,3,5], p = 2, q = 8
Output: 6
Explanation: The LCA of nodes 2 and 8 is 6.
```
Example 2:

```
Input: root = [6,2,8,0,4,7,9,null,null,3,5], p = 2, q = 4
Output: 2
Explanation: The LCA of nodes 2 and 4 is 2, since a node can be a descendant of itself according to the LCA definition.
```
Example 3:

```
Input: root = [2,1], p = 2, q = 1
Output: 2
```
 
Constraints:

* The number of nodes in the tree is in the range [2, 105].
* -109 <= Node.val <= 109
* All Node.val are unique.
* p != q
* p and q will exist in the BST.

Big O

* if use recursion, time is $O(N)$, space is $O(N)$
* if do not use recursion, time is $O(logn)$, space is $O(1)$

**Key Takeaway**
* binary tree property based on the comparing p, q value to the current root value
* The base case is current root value

In [91]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # Interative
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        cur = root
        while cur is not None:
            if p.val > cur.val and q.val > cur.val:
                cur = cur.right
            elif p.val < cur.val and q.val < cur.val:
                cur = cur.left
            else:
                return cur

    # recursive
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # Value of current node or parent node.
        parent_val = root.val

        # Value of p
        p_val = p.val

        # Value of q
        q_val = q.val

        # If both p and q are greater than parent
        if p_val > parent_val and q_val > parent_val:    
            return self.lowestCommonAncestor(root.right, p, q)
        # If both p and q are lesser than parent
        elif p_val < parent_val and q_val < parent_val:    
            return self.lowestCommonAncestor(root.left, p, q)
        # We have found the split point, i.e. the LCA node.
        else:
            return root

## (110) Balanced Binary Tree

Given a binary tree, determine if it is height-balanced

Example 1:

```
Input: root = [3,9,20,null,null,15,7]
Output: true
```
Example 2:

```
Input: root = [1,2,2,3,3,null,null,4,4]
Output: false
```
Example 3:
```
Input: root = []
Output: true
```

Constraints:

* The number of nodes in the tree is in the range [0, 5000].
* -104 <= Node.val <= 104

In [ ]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def height(self, root):
        if root is None:
            return -1
        return 1 + max(height(root.left), height(root.right))
        
        
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        if abs(self.height(root.left) - self.height(root.right))>1:
            return False
        else:
            return self.isBalanced(root.left) && self.isBalanced(root.right)

class Solution:
    # Return whether or not the tree at root is balanced while also returning
    # the tree's height
    def isBalancedHelper(self, root: TreeNode) -> (bool, int):
        # An empty tree is balanced and has height -1
        if not root:
            return True, -1
        
        # Check subtrees to see if they are balanced. 
        leftIsBalanced, leftHeight = self.isBalancedHelper(root.left)
        if not leftIsBalanced:
            return False, 0
        rightIsBalanced, rightHeight = self.isBalancedHelper(root.right)
        if not rightIsBalanced:
            return False, 0
        
        # If the subtrees are balanced, check if the current tree is balanced
        # using their height
        return (abs(leftHeight - rightHeight) < 2), 1 + max(leftHeight, rightHeight)
        
    def isBalanced(self, root: TreeNode) -> bool:
        return self.isBalancedHelper(root)[0]

## (141) Linked List Cycle

Given head, the head of a linked list, determine if the linked list has a cycle in it.

There is a cycle in a linked list if there is some node in the list that can be reached again by continuously following the next pointer. Internally, pos is used to denote the index of the node that tail's next pointer is connected to. Note that pos is not passed as a parameter.

Return true if there is a cycle in the linked list. Otherwise, return false.

 

Example 1:

```
Input: head = [3,2,0,-4], pos = 1
Output: true
Explanation: There is a cycle in the linked list, where the tail connects to the 1st node (0-indexed).
```
Example 2:

```
Input: head = [1,2], pos = 0
Output: true
Explanation: There is a cycle in the linked list, where the tail connects to the 0th node.
```
Example 3:

```
Input: head = [1], pos = -1
Output: false
Explanation: There is no cycle in the linked list.
```

Constraints:

* The number of the nodes in the list is in the range [0, 104].
* -105 <= Node.val <= 105
* pos is -1 or a valid index in the linked-list.


Big O
* Time: $O(n)$
* Space $O(1)$ if we use 2 pointer (fast, slow) $O(n)$ if we use hashmap

**Key takeaway**
* use a fast and slow pointer, fast is 2x faster than slow, this utilize Floyd's cycle-finding algorithm, also known as the "tortoise and hare" algorithm, to determine the presence of a loop efficiently.
* If there is a loop in the list, the fast pointer will eventually meet the slow pointer.
* Or use a hashmap if it is there in the hashmap, we find the cycle

In [ ]:
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False
        
        cur = head
        hashmap = {}
        while cur is not None:
            hashmap[cur] = 1+hashmap.get(cur, 0)
            cur = cur.next
            if hashmap.get(cur, 0) > 1:
                return True
        return False
    
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False
        
        slow, fast = head, head
        while fast is not None and fast.next is not None:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                return True
        return False

## (232) Implement Queue Using Stacks

Implement a first in first out (FIFO) queue using only two stacks. The implemented queue should support all the functions of a normal queue (push, peek, pop, and empty).

Implement the MyQueue class:

* `void push(int x)` Pushes element x to the back of the queue.
* `int pop()` Removes the element from the front of the queue and returns it.
* `int peek()` Returns the element at the front of the queue.
* `boolean empty()` Returns true if the queue is empty, false otherwise.

Notes:

* You must use only standard operations of a stack, which means only push to top, peek/pop from top, size, and is empty operations are valid.
* Depending on your language, the stack may not be supported natively. You may simulate a stack using a list or deque (double-ended queue) as long as you use only a stack's standard operations.
 

Example 1:

```
Input
["MyQueue", "push", "push", "peek", "pop", "empty"]
[[], [1], [2], [], [], []]
Output
[null, null, null, 1, 1, false]
```

Explanation
```
MyQueue myQueue = new MyQueue();
myQueue.push(1); // queue is: [1]
myQueue.push(2); // queue is: [1, 2] (leftmost is front of the queue)
myQueue.peek(); // return 1
myQueue.pop(); // return 1, queue is [2]
myQueue.empty(); // return false
```
 

Constraints:

* 1 <= x <= 9
* At most 100 calls will be made to push, pop, peek, and empty.
* All the calls to pop and peek are valid.

Follow-up: Can you implement the queue such that each operation is amortized O(1) time complexity? In other words, performing n operations will take overall O(n) time even if one of those operations may take longer.

Big O
* Time: $O(n)$ for push or for pop based on approach

**key takeaway**

use the second stack to order or the queue when adding a new element

In [117]:
class MyQueue: # push O(n), pop O(1)

    def __init__(self):
        self.stack1 = []
        self.stack2 = []

    def push(self, x: int) -> None:
        while len(self.stack1) > 0:
            self.stack2.append(self.stack1.pop(-1))
         
        self.stack1.append(x)
        
        while len(self.stack2) > 0:
            self.stack1.append(self.stack2.pop(-1))
        

    def pop(self) -> int:
        if len(self.stack1) == 0:
            return None
        return self.stack1.pop(-1)

    def peek(self) -> int:
        if len(self.stack1) == 0:
            return None
        return self.stack1[-1]

    def empty(self) -> bool:
        return len(self.stack1) == 0

class MyQueue: # push O(1), pop O(n)
    def __init__(self):
        self.s1 = []
        self.s2 = []

    def push(self, x):
        self.s1.append(x)

    def pop(self):
        self.peek()
        return self.s2.pop()

    def peek(self):
        if not self.s2:
            while self.s1:
                self.s2.append(self.s1.pop())
        return self.s2[-1]        

    def empty(self):
        return not self.s1 and not self.s2

# Week 2

## (278) First Bad Version

You are a product manager and currently leading a team to develop a new product. Unfortunately, the latest version of your product fails the quality check. Since each version is developed based on the previous version, all the versions after a bad version are also bad.

Suppose you have n versions [1, 2, ..., n] and you want to find out the first bad one, which causes all the following ones to be bad.

You are given an API bool isBadVersion(version) which returns whether version is bad. Implement a function to find the first bad version. You should minimize the number of calls to the API.

 

Example 1:
```
Input: n = 5, bad = 4
Output: 4
Explanation:
call isBadVersion(3) -> false
call isBadVersion(5) -> true
call isBadVersion(4) -> true
Then 4 is the first bad version.
```
Example 2:
```
Input: n = 1, bad = 1
Output: 1
```

Constraints:

1 <= bad <= n <= 2^31 - 1

Big O:
time: $O(logn)$
space: $O(1)$

**key takeaway**

Use binary search to find with minimum call, do not add 1 to right (high) when find it a bad version because it needs to be in the next search space in case it is the last one.

In [ ]:
# The isBadVersion API is already defined for you.
def isBadVersion(version: int) -> bool:
    pass

class Solution:
    def firstBadVersion(self, n: int) -> int:
        if n == 1:
            return 1 if isBadVersion(1) else None
        low = 1
        high = n
        while low < high:
            mid = int((high + low)/2)
            if isBadVersion(mid):
                high = mid
            else:
                low = mid + 1
        return low

## (383) Ransom Note

Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.

Each letter in magazine can only be used once in ransomNote.


Example 1:
```
Input: ransomNote = "a", magazine = "b"
Output: false
```
Example 2:
```
Input: ransomNote = "aa", magazine = "ab"
Output: false
```
Example 3:
```
Input: ransomNote = "aa", magazine = "aab"
Output: true
```

Constraints:

* 1 <= ransomNote.length, magazine.length <= 105
* ransomNote and magazine consist of lowercase English letters.

Big O
* Time: $O(m)$
* Space $O(k)$ where k is the 26 (alphabets)

**key Takeaway**
Using hashmap to store available characters and counts, iterate ransomNote to decrease count if it depletes the characters store, it is false, otherwise true

In [3]:
import collections

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        if len(magazine) < len(ransomNote):
            return False
        
        
        hashmap = {}
        for c in magazine:
            hashmap[c] = 1 + hashmap.get(c, 0)
        
        for c in ransomNote:
            if c not in hashmap:
                return False
            else:
                hashmap[c] -= 1
                if hashmap[c] < 0:
                    return False
        return True

    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        if len(magazine) < len(ransomNote):
            return False
        
        
        letters = collections.Counter(magazine)
        for c in ransomNote:
            # If there are none of c left, return False.
            if letters[c] <= 0:
                return False
            # Remove one of c from the Counter.
            letters[c] -= 1
        return True
    
ransomNote = "chejaccdae"
magazine = "geceeibccchjejhdd"
s = Solution()
s.canConstruct(ransomNote, magazine)

False

## (70) Climbing Stairs

You are climbing a staircase. It takes n steps to reach the top.

Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

 
Example 1:
```
Input: n = 2
Output: 2
Explanation: There are two ways to climb to the top.
1. 1 step + 1 step
2. 2 steps
```
Example 2:
```
Input: n = 3
Output: 3
Explanation: There are three ways to climb to the top.
1. 1 step + 1 step + 1 step
2. 1 step + 2 steps
3. 2 steps + 1 step
``` 

Constraints:

* 1 <= n <= 45

Big O

* Time: $O(2^n)$ if use recursion too slow. $O(n) if we use dynamic programming


**key takeaway**
Use bottom-up dynamic programming

In [2]:
class Solution:
    def __init__(self):
        self.cache = {1:1, 2:2}

    # pure recursive
    def climbStairs_recursion(self, n: int) -> int:
        if n == 1:
            return 1
        elif n == 2:
            return 2
        else:
            return self.climbStairs_recursion(n-1) + self.climbStairs_recursion(n-2)
    
    # recursion with hashtable
    def climbStairs_cache(self, n: int) -> int:
        if n == 1:
            return 1
        if n == 2:
            return 2
        
        if n in self.cache:
            return self.cache[n]
        else:
            ways = self.climbStairs_cache(n-1) + self.climbStairs_cache(n-2)
            self.cache[n] = ways
            return ways
        
    def climbStairs_dp(self, n: int) -> int:
        if n == 1:
            return 1
        if n == 2:
            return 2
        
        N_2, N_1 = 1, 2
        for N in range(3, n+1):
            temp = N_1
            Next = N_2 + N_1
            N_1 = Next
            N_2 = temp
        return Next

s = Solution()
#%timeit -n5 s.climbStairs_recursion(30)
%timeit -n10 s.climbStairs_cache(100)
%timeit -n10 s.climbStairs_dp(100)

The slowest run took 26.96 times longer than the fastest. This could mean that an intermediate result is being cached.
2.23 µs ± 4.25 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
6.86 µs ± 1.18 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


## (409) Longest Palindrome

Given a string s which consists of lowercase or uppercase letters, return the length of the longest palindrome that can be built with those letters.

Letters are case sensitive, for example, "Aa" is not considered a palindrome here.

 

Example 1:
```
Input: s = "abccccdd"
Output: 7
Explanation: One longest palindrome that can be built is "dccaccd", whose length is 7.
```
Example 2:
```
Input: s = "a"
Output: 1
Explanation: The longest palindrome that can be built is "a", whose length is 1.
```

Constraints:

* 1 <= s.length <= 2000
* s consists of lowercase and/or uppercase English letters only.

Big O

* Time O(n)
* Space O(1)

**Key takeaway**

Use a hashtable to store character frequency, determine whether there is a unique center or not if yes need to add 1

In [9]:
class Solution:
    def longestPalindrome(self, s: str) -> int:
        store = collections.Counter(s)
        even_size = 0
        odd_size = 0
        odd_count = 0
        for c, v in store.items():
            if v%2 == 0:
                even_size += v
            else:
                odd_size += v - 1
                odd_count+=1
        if odd_count == 0:
            return even_size
        else:
            return even_size + odd_size + 1        

## (206) Reverse Linked List

Given the head of a singly linked list, reverse the list, and return the reversed list.

 

Example 1:

```
Input: head = [1,2,3,4,5]
Output: [5,4,3,2,1]
```
Example 2:

```
Input: head = [1,2]
Output: [2,1]
```
Example 3:
```
Input: head = []
Output: []
``` 

Constraints:

* The number of nodes in the list is the range [0, 5000].
* -5000 <= Node.val <= 5000

Big O
* time $O(n)$
* mem $O(1)$

**Key takeaway**

create current (set to head) and prev (set to Null), 2 pointers, reverse the link and shift 2 pointer forward until the current read the end (become none), return prev since it is points to the last node

In [ ]:
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr = head
        prev = None
        while curr is not None:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        return prev

## (169) Majority Element

Given an array nums of size n, return the majority element.

The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority element always exists in the array.

 

Example 1:
```
Input: nums = [3,2,3]
Output: 3
```
Example 2:
```
Input: nums = [2,2,1,1,1,2,2]
Output: 2
```

Constraints:

* n == nums.length
* 1 <= n <= 5 * 104
* -109 <= nums[i] <= 109

Big O
if use hashmap

* Time $O(n)$
* Space $O(n)$

if use sort

* Time $O(nlogn)$
* Space $O(1)$

**Key Takeaway**

* hashmap is straightfoward but more costly
* sort method: sort the list first, the majority element will always occupy the middle 1 or 2 index depending on the length
* **Boyer-Moore Voting Algorithm** is a the best approach
    * Use the first element as the candidate, keep a counter
    * iterate throw the list, if the next element is the same element +1, if different -1
    * when the counter = 0, the next element becomes the new candidate
    * In the end of iteration, the candidate will be the majority element

In [22]:
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        freq_counter = collections.Counter(nums)
        for k, v in freq_counter.items():
            if v == max(freq_counter.values()):
                return k
            
    def majorityElement(self, nums: List[int]) -> int:
        count = {}
        res, maxCount = 0, 0
        for n in nums:
            count[n] = 1 + count.get(n, 0)
            res = n if count[n] > maxCount else res
            maxCount = max(count[n], maxCount)
        return res
    
    def majorityElement(self, nums: List[int]) -> int:
        freq_counter = collections.Counter(nums)
        return max(freq_counter.keys(), key=counts.get)
    
    def majorityElement(self, nums: List[int]) -> int:
        majority_count = len(nums) // 2 + 1
        freq_counter = collections.Counter(nums)
        for k, v in freq_counter.items():
            if v >= majority_count:
                return k
    
    def majorityElement(self, nums):
        nums.sort()
        return nums[len(nums)//2]

    # Boyer-moore algorithm
    def majorityElement(self, nums: List[int]) -> int:
        count, candidate = 0, 0
        for n in nums:
            if count == 0:
                candidate = n
            count += (1 if n == candidate else - 1)
        return candidate
    

nums = [2, 2, 1, 1, 1, 2, 2]
s = Solution()
s.majorityElement(nums)

2

## (67) Add Binary

Given two binary strings a and b, return their sum as a binary string.


Example 1:
```
Input: a = "11", b = "1"
Output: "100"
```
Example 2:
```
Input: a = "1010", b = "1011"
Output: "10101"
```
 
Constraints:

* 1 <= a.length, b.length <= 104
* a and b consist only of '0' or '1' characters.
* Each string does not contain leading zeros except for the zero itself.

Big O

* Time $O(n)$
* Space $O(1)$

**key takeaway**

Similar to other add problem, keep a carry bits, reverse order or iterate reversely, use `ord()` function to calculate number using ascii values, total%2 naturally gives the char, total//2 naturally gives the carry

In [39]:
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        result = ''
        carry = 0
        max_len = max(len(a), len(b))
        a = a.zfill(max_len)
        b = b.zfill(max_len)
        
        for i in range(max_len - 1, -1 ,-1):
            num1 = 1 if a[i] == '1' else 0
            num2 = 1 if b[i] == '1' else 0
            result = ('1' if (num1 + num2 + carry) % 2 == 1 else '0') + result
            carry = 0 if num1 + num2 + carry < 2 else 1
        if carry != 0:
            result = '1' + result
        return result
    
    def addBinary(self, a: str, b: str) -> str:
        res = ''
        carry = 0
        a, b = a[::-1], b[::-1]
        for i in range(max(len(a), len(b))):
            digit_a = ord(a[i]) - ord("0") if i < len(a) else 0
            digit_b = ord(b[i]) - ord("0") if i < len(b) else 0
            
            total = digit_a + digit_b + carry
            char = total%2
            res = char + res
            carry = total//2
        if carry != 0:
            res = '1' + res
        return res
            
    # using bit manipulation
    # XOR is sum without carray
    # AND and shift 1 bits to the left find the carray
    # we keep shift carry to the left unit it becomes 0
    def addBinary(self, a, b) -> str:
        x, y = int(a, 2), int(b, 2)
        while y:
            answer = x ^ y
            carry = (x & y) << 1
            x, y = answer, carry
        return bin(x)[2:]
    

s = Solution()
s.addBinary(a='11', b = '1')
s.addBinary(a='1010', b = '101100')

'110110'

## (543) Diameter of Binary Tree

Given the root of a binary tree, return the length of the diameter of the tree.

The diameter of a binary tree is the length of the longest path between any two nodes in a tree. This path may or may not pass through the root.

The length of a path between two nodes is represented by the number of edges between them.

 

Example 1:

```
Input: root = [1,2,3,4,5]
Output: 3
Explanation: 3 is the length of the path [4,2,1,3] or [5,2,1,3].
```
Example 2:
```
Input: root = [1,2]
Output: 1
```
 

Constraints:

* The number of nodes in the tree is in the range [1, 104].
* -100 <= Node.val <= 100

Big O

* Time: $O(n)$
* Space $O(n)$

**key takeway**
* DFS
* initalize an integer variable diameter to keep track of the longest path we find from the DFS.
* Implement a recursive function longestPath which takes a TreeNode as input. It should recursively explore the entire tree rooted at the given node. Once it's finished, it should return the longest path out of its left and right branches:
* if node is None, we have reached the end of the tree, hence we should return 0;
* we want to recursively explore node's children, so we call longestPath again with node's left and right children. In return, we get the longest path of its left and right children leftPath and rightPath;
* if leftPath plus rightPath is longer than the current longest diameter found, then we need to update diameter;
* finally, we return the longer one of leftPath and rightPath. Remember to add 111 as the edge connecting it with its parent.
* Call longestPath with root.

In [ ]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter = 0
        
        def traverse(node):
            if not node:
                return 0
            nonlocal diameter
            left = traverse(node.left)
            right = traverse(node.right)
           
            diameter = max(diameter, left + right)
            return max(left, right) + 1
        
        traverse(root)
        return diameter
            


s = Solution()
s.diameterOfBinaryTree

## (876) Middle of the Linked List

Given the head of a singly linked list, return the middle node of the linked list.

If there are two middle nodes, return the second middle node.

 

Example 1:

```
Input: head = [1,2,3,4,5]
Output: [3,4,5]
Explanation: The middle node of the list is node 3.
```
Example 2:

```
Input: head = [1,2,3,4,5,6]
Output: [4,5,6]
Explanation: Since the list has two middle nodes with values 3 and 4, we return the second one.
```

Constraints:

* The number of nodes in the list is in the range [1, 100].
* 1 <= Node.val <= 100

Big O

* Time: $O(n)$
* Space: $O(1)$

**key takeaway**

Use 2 pointer, slow and fast, fast go 2x as the slow, when fast reach the end, slow reach the middle

In [ ]:
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = head
        fast = head
        while fast is not None and fast.next is not None:
            fast = fast.next.next
            slow = slow.next
        
        return flow

## (104) Maximum Depth of Binary Tree

Given the root of a binary tree, return its maximum depth.

A binary tree's maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.

 

Example 1:

```
Input: root = [3,9,20,null,null,15,7]
Output: 3
```
Example 2:
```
Input: root = [1,null,2]
Output: 2
```
 

Constraints:

* The number of nodes in the tree is in the range [0, 104].
* -100 <= Node.val <= 100

Big O

* Time $O(logn)$
* space $O(logn)$

In [42]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        left_size, right_size = 0, 0
        if root.left is not None:
            left_size = self.maxDepth(node.left)
        if node.right is not None:
            right_size = self.maxDepth(node.right)
        return 1 + max(left_size, right_size)

## (217) Contains Duplicate

Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.

 

Example 1:
```
Input: nums = [1,2,3,1]
Output: true
```
Example 2:
```
Input: nums = [1,2,3,4]
Output: false
```
Example 3:
```
Input: nums = [1,1,1,3,3,4,3,2,4,2]
Output: true
```
 

Constraints:

* 1 <= nums.length <= 105
* -109 <= nums[i] <= 109

Big O

* Time $O(nlogn)$ (use sorting)
* Space $O(1)$

**key takeaway**

We can use set but it may not be optimal. We can either use sorted first then linear search or use hashtable

In [44]:
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        return len(set(nums)) != len(nums)
    
    def containsDuplicate(self, nums: List[int]) -> bool:
        nums = sorted(nums)
        for i in range(1, len(nums)):
            if nums[i-1] == nums[i]:
                return True
        return False

## (53) Maximum Subarray -- (Kadane's Algorithm)

Given an integer array nums, find the subarray with the largest sum, and return its sum.

 

Example 1:
```
Input: nums = [-2,1,-3,4,-1,2,1,-5,4]
Output: 6
Explanation: The subarray [4,-1,2,1] has the largest sum 6.
```
Example 2:
```
Input: nums = [1]
Output: 1
Explanation: The subarray [1] has the largest sum 1.
```
Example 3:
```
Input: nums = [5,4,-1,7,8]
Output: 23
Explanation: The subarray [5,4,-1,7,8] has the largest sum 23.
```
 
Constraints:

* 1 <= nums.length <= 10^5
* -10^4 <= nums[i] <= 10^4
 
Big O

* Time $O(n)$
* Space $O(1)$

**Key Takeaway**

* If we use brunt force, we need to iterate every combination and perform sum calculate => $O(n^3)$
* We can use cumulative sum to bring the complexity to $O(n^2)$ but still very slow
* Use dynamic programming to reduce $O(n)$
    * Current max is max(previous max + current number, current number)

**Whenever you see a question that asks for the maximum or minimum of something, consider Dynamic Programming as a possibility**

In [108]:
class Solution:
    # Brute Force with cumSum: Time limit exceed with O(n^2)
    def maxSubArray(self, nums: List[int]) -> int:
        max_val = nums[0]
        for i in range(len(nums)):
            curSum = 0
            for j in range(i, len(nums)):
                curSum += nums[j]
                max_val = max(max_val, curSum)
        return max_val
    
    # Dynamic programming
    def maxSubArray(self, nums: List[int]) -> int:
        max_val = [nums[0]]
        for i in range(1, len(nums)):
            max_val.append(max(max_val[i-1] + nums[i], nums[i]))
            #max_val.append(max(max_val[i-1], 0) + nums[i])
    
        return max(max_val)

    def maxSubArray(self, nums: List[int]) -> int:
        max_val = nums[0]
        curSum = nums[0]
        for i in range(1, len(nums)):
            curSum = max(curSum + nums[i], nums[i])
            #curSum = max(curSum, 0) + nums[i]
            max_val = max(max_val, curSum)
    
        return max_val
            
                      
s = Solution()
s.maxSubArray([5, 4, -1, 7, 8])
#s.maxSubArray([-2,1,-3,4,-1,2,1,-5,4])
#s.maxSubArray([-2,-3,-1])
#s.maxSubArray([-1,0,-2])

23

# Week 3

## (57) Insert Interval

You are given an array of non-overlapping intervals intervals where intervals[i] = [starti, endi] represent the start and the end of the ith interval and intervals is sorted in ascending order by starti. You are also given an interval newInterval = [start, end] that represents the start and end of another interval.

Insert newInterval into intervals such that intervals is still sorted in ascending order by starti and intervals still does not have any overlapping intervals (merge overlapping intervals if necessary).

Return intervals after the insertion.

 

Example 1:
```
Input: intervals = [[1,3],[6,9]], newInterval = [2,5]
Output: [[1,5],[6,9]]
```
Example 2:
```
Input: intervals = [[1,2],[3,5],[6,7],[8,10],[12,16]], newInterval = [4,8]
Output: [[1,2],[3,10],[12,16]]
Explanation: Because the new interval [4,8] overlaps with [3,5],[6,7],[8,10].
``` 

Constraints:

* 0 <= intervals.length <= 104
* intervals[i].length == 2
* 0 <= starti <= endi <= 105
* intervals is sorted by starti in ascending order.
* newInterval.length == 2
* 0 <= start <= end <= 105

In [118]:
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []
        for i in range(len(intervals)):
            # edge case 1: new interval is not overlapping with the current interval
            # and it is smaller then the first element in the interval
            if newInterval[1] < intervals[i][0]:
                res.append(newInterval)
                return res + intervals[i:]
            # new interval is not overlapping with current interval
            # it may overlapping with next one, we append the current interval
            elif newInterval[0] > intervals[i][1]:
                res.append(intervals[i])
            # there is overlapping
            else:
                newInterval = [min(newInterval[0], intervals[i][0]), max(newInterval[1], intervals[i][1])]
        res.append(newInterval) 
        return res
            

s = Solution()
s.insert([[1,3],[6,9]], [2, 5])
#s.insert([[1,2],[3,5],[6,7],[8,10],[12,16]], [4, 8])

[[1, 5], [6, 9]]

## (542) 01 Matrix

Given an m x n binary matrix mat, return the distance of the nearest 0 for each cell.

The distance between two adjacent cells is 1.

 

Example 1:

```
Input: mat = [[0,0,0],[0,1,0],[0,0,0]]
Output: [[0,0,0],[0,1,0],[0,0,0]]
```
Example 2:

```
Input: mat = [[0,0,0],[0,1,0],[1,1,1]]
Output: [[0,0,0],[0,1,0],[1,2,1]]
```

Constraints:

* m == mat.length
* n == mat[i].length
* 1 <= m, n <= 104
* 1 <= m * n <= 104
* mat[i][j] is either 0 or 1.
* There is at least one 0 in mat.


**key takeaway**

Finding the nearst ... is usually use BFS, to improve efficiency, it is not aways search from the source to target but also search from the target to the source

**Create a BFS Template to use**

In [148]:
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        def valid(row, col):
            return 0 <= row < m and 0 <= col < n
        
        matrix = [row[:] for row in mat]
        m = len(matrix)
        n = len(matrix[0])
        queue = deque()
        seen = set()
        
        for row in range(m):
            for col in range(n):
                if matrix[row][col] == 0:
                    queue.append((row, col, 0))
                    seen.add((row, col))
        
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]

        while queue:
            row, col, steps = queue.popleft()
            
            for dx, dy in directions:
                next_row, next_col = row + dy, col + dx
                if (next_row, next_col) not in seen and valid(next_row, next_col):
                    seen.add((next_row, next_col))
                    queue.append((next_row, next_col, steps + 1))
                    matrix[next_row][next_col] = steps + 1
        return matrix

## K Closest Points to Origin

Given an array of points where points[i] = [xi, yi] represents a point on the X-Y plane and an integer k, return the k closest points to the origin (0, 0).

The distance between two points on the X-Y plane is the Euclidean distance (i.e., √(x1 - x2)2 + (y1 - y2)2).

You may return the answer in any order. The answer is guaranteed to be unique (except for the order that it is in).

```
Input: points = [[1,3],[-2,2]], k = 1
Output: [[-2,2]]
Explanation:
The distance between (1, 3) and the origin is sqrt(10).
The distance between (-2, 2) and the origin is sqrt(8).
Since sqrt(8) < sqrt(10), (-2, 2) is closer to the origin.
We only want the closest k = 1 points from the origin, so the answer is just [[-2,2]].
```

Example 2:
```
Input: points = [[3,3],[5,-1],[-2,4]], k = 2
Output: [[3,3],[-2,4]]
Explanation: The answer [[-2,4],[3,3]] would also be accepted.
```

Constraints:

* 1 <= k <= points.length <= 104
* -104 < xi, yi < 104

Big O:
* Time $O(n)$ if use heap
* Time $O(nlogn)$ if use sort approach
* Space $O(n)$

**Key takeaway**

* Use min heap (python provided data structure) to store [distance, x, y] of each points, the pop k ties
* Just sort the points direction using a lambda function

In [183]:
import heapq

class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:

        def euc_dist(x, y):
            return x**2 + y**2
        
        minHeap = []
        for x, y in points:
            minHeap.append([euc_dist(x, y), x, y])
        heapq.heapify(minHeap)
        res = []
        while k > 0:
            dist, x, y = heapq.heappop(minHeap)
            res.append([x, y])
            k-=1
        return res

    def kClosest(self, points, K):
        points.sort(key = lambda P: P[0]**2 + P[1]**2)
        return points[:K]
    
points = [[2,2],[2,2],[3,3],[2,-2],[1,1]]
k = 4
s = Solution()
s.kClosest(points, k)

[[1, 1], [2, -2], [2, 2], [2, 2]]

## (3)  Longest Substring Without Repeating Characters

Given a string s, find the length of the longest 
substring
 without repeating characters.

 

Example 1:
```
Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.
```
Example 2:
```
Input: s = "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.
```
Example 3:
```
Input: s = "pwwkew"
Output: 3
Explanation: The answer is "wke", with the length of 3.
Notice that the answer must be a substring, "pwke" is a subsequence and not a substring.
```

Constraints:

* 0 <= s.length <= 5 * 104
* s consists of English letters, digits, symbols and spaces.

Big O

* Time $O(n)$
* space $O(n)$

**key takeaway**

use **Sliding Window** method, 2 pointers left and right, if right pointer find duplicates, shift left pointer until it not longer contain duplicates

In [132]:
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        charSet = set()
        l = 0
        res = 0
        for r in range(len(s)): # iterate right pointer
            # if it is repeat move left forward
            while s[r] in charSet:
                charSet.remove(s[l])
                l+=1
            charSet.add(s[r])
            res = max(res, r - l + 1)
        return res

s = Solution()
#s.lengthOfLongestSubstring("abcabcbb")
#s.lengthOfLongestSubstring("bbbbb")
#s.lengthOfLongestSubstring("pwwkew")
s.lengthOfLongestSubstring("dvdf")

3

## (167) Two Sum II - Input Array is Sorted

Given a 1-indexed array of integers numbers that is already sorted in non-decreasing order, find two numbers such that they add up to a specific target number. Let these two numbers be `numbers[index1]` and `numbers[index2]` where `1 <= index1 < index2 < numbers.length`.

Return the indices of the two numbers, index1 and index2, added by one as an integer array [index1, index2] of length 2.

The tests are generated such that there is exactly one solution. You may not use the same element twice.

Your solution must use only constant extra space.

 

Example 1:
```
Input: numbers = [2,7,11,15], target = 9
Output: [1,2]
Explanation: The sum of 2 and 7 is 9. Therefore, index1 = 1, index2 = 2. We return [1, 2].
```
Example 2:
```
Input: numbers = [2,3,4], target = 6
Output: [1,3]
Explanation: The sum of 2 and 4 is 6. Therefore index1 = 1, index2 = 3. We return [1, 3].
```
Example 3:
```
Input: numbers = [-1,0], target = -1
Output: [1,2]
Explanation: The sum of -1 and 0 is -1. Therefore index1 = 1, index2 = 2. We return [1, 2].
``` 

Constraints:

* 2 <= numbers.length <= 3 * 104
* -1000 <= numbers[i] <= 1000
* numbers is sorted in non-decreasing order.
* -1000 <= target <= 1000
* The tests are generated such that there is exactly one solution.

Big O
* Time complexity $O(N)$
* Space complexity $O(1)$

**key takeaway**
* User 2 pointers, one from the low end, one from the high end, if sum is < target, increase low index, if num is > target decrease high index

In [189]:
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        i = 0
        j = len(numbers) - 1
        while i < len(numbers) and j >= 0:
            if numbers[i] + numbers[j] < target:
                i+=1
            elif numbers[i] + numbers[j] > target:
                j-=1
            else:
                return [i+1, j+1]
            
numbers = [2,7,11,15]
target = 9
s = Solution()
s.twoSum(numbers, target)

[1, 2]

## (15) 3Sum

Given an integer array nums, return all the triplets `[nums[i], nums[j], nums[k]]` such that `i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.`

Notice that the solution set must not contain duplicate triplets.
 

Example 1:
```
Input: nums = [-1,0,1,2,-1,-4]
Output: [[-1,-1,2],[-1,0,1]]
Explanation: 
nums[0] + nums[1] + nums[2] = (-1) + 0 + 1 = 0.
nums[1] + nums[2] + nums[4] = 0 + 1 + (-1) = 0.
nums[0] + nums[3] + nums[4] = (-1) + 2 + (-1) = 0.
The distinct triplets are [-1,0,1] and [-1,-1,2].
Notice that the order of the output and the order of the triplets does not matter.
```
Example 2:
```
Input: nums = [0,1,1]
Output: []
Explanation: The only possible triplet does not sum up to 0.
```
Example 3:
```
Input: nums = [0,0,0]
Output: [[0,0,0]]
Explanation: The only possible triplet sums up to 0.
``` 

Constraints:

* 3 <= nums.length <= 3000
* -105 <= nums[i] <= 105

Big O:
* Time $O(nlogn) + O(n^2)$ ==> $O(n^2)$
* Space $O(n)$

In [234]:
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) == 3:
            if sum(nums) == 0:
                return [nums]
            else:
                return []
        
        res = []
        sorted_nums = sorted(nums)
        for i in range(len(sorted_nums)-2):
            num = sorted_nums[i]
            if i > 0 and num == sorted_nums[i-1]:
                continue

            l = i+1
            r = len(sorted_nums)-1
            while l < r:
                threesum = num + sorted_nums[l] + sorted_nums[r]
                if threesum > 0:
                    r-=1
                elif threesum <0:
                    l+=1
                else:
                    res.append([num, sorted_nums[l], sorted_nums[r]])
                    l+=1
                    while nums[l] == nums[1 - 1] and l < r:
                        l+=1
        return res

nums = [-2,0,1,1,2]
s = Solution()
s.threeSum(nums)

[[-2, 0, 2], [-2, 1, 1]]